File: csharp/806.number-of-lines-to-write-string.cs
/*
 * @lc app=leetcode id=806 lang=csharp
 *
 * [806] Number of Lines To Write String
 */

using System.Linq;

// @lc code=start
public class Solution {
    public int[] NumberOfLines(int[] widths, string S) {
        return S
            .Select(c => widths[(int)c - (int)'a'])
            .Aggregate(
                (Line: 1, Count: 0), 
                (tuple, length) => 
                    length + tuple.Count > 100 ? 
                    (Line: tuple.Line + 1, Count: length) : 
                    (Line: tuple.Line, Count: tuple.Count + length), 
                tuple => new int[]{ tuple.Line, tuple.Count });
    }
}
// @lc code=end

File: java/806.number-of-lines-to-write-string.java
/*
 * @lc app=leetcode id=806 lang=java
 *
 * [806] Number of Lines To Write String
 */

// @lc code=start
class Solution {
    public int[] numberOfLines(int[] widths, String s) {
        int acc = 0, lines = 0;
        for (int i = 0; i < s.length(); ++i) {
            int w = widths[(int)s.charAt(i) - (int)'a'];
            if (acc + w > 100) {
                ++lines;
                acc = w;
            } else {
                acc += w;
            }
        }
        return new int[] { lines + 1, acc };
    }
}
// @lc code=end

